kvbc/test/categorization/blocks_test.cpp

```diff
@@ -133,6 +158,183 @@ TEST_F(categorized_kvbc, fail_reconstruct_merkle_updates) {
+
+    ASSERT_EQ(block_chain.addBlock(std::move(updates)), (BlockId)2);
+  }
+
+  {
+    const detail::Buffer& block_db_key = Block::generateKey(1);
+    auto block1_db_val = db->get(BLOCKS_CF, block_db_key);
+    ASSERT_TRUE(block1_db_val.has_value());
+
+    detail::Buffer in{block1_db_val.value().begin(), block1_db_val.value().end()};
+    auto block1_from_db = Block::deserialize(in);
+
+    categorization::RawBlock rw(block1_from_db, db);
+    auto variant = rw.data.updates.kv["imm"];
+    auto imm_updates = std::get<ImmutableInput>(variant);
+    // check reconstruction of original kv
+    ASSERT_EQ(imm_updates.kv.size(), 2);
+    auto ex1 = ImmutableValueUpdate{"imm_value1", {"1", "2"}};
+    auto ex2 = ImmutableValueUpdate{"imm_value2", {"12", "22"}};
+    ASSERT_EQ(imm_updates.kv["imm_key1"], ex1);
+    ASSERT_EQ(imm_updates.kv["imm_key2"], ex2);
+  }
+
+  {
+    const detail::Buffer& block_db_key = Block::generateKey(2);
+    auto block2_db_val = db->get(BLOCKS_CF, block_db_key);
+    ASSERT_TRUE(block2_db_val.has_value());
+
+    detail::Buffer in{block2_db_val.value().begin(), block2_db_val.value().end()};
+    auto block2_from_db = Block::deserialize(in);
+
+    categorization::RawBlock rw(block2_from_db, db);
+    auto variant = rw.data.updates.kv["imm"];
+    auto imm_updates = std::get<ImmutableInput>(variant);
+    // check reconstruction of original kv
+    ASSERT_EQ(imm_updates.kv.size(), 1);
+    auto ex1 = ImmutableValueUpdate{"imm_value3", {}};
+    ASSERT_EQ(imm_updates.kv["imm_key3"], ex1);
+
+    auto variant2 = rw.data.updates.kv["imm2"];
+    auto imm_updates2 = std::get<ImmutableInput>(variant2);
+    // check reconstruction of original kv
+    ASSERT_EQ(imm_updates2.kv.size(), 1);
+    auto ex2 = ImmutableValueUpdate{"imm_value32", {"32"}};
+    ASSERT_EQ(imm_updates2.kv["imm_key12"], ex2);
+  }
+}
+
+TEST_F(categorized_kvbc, fail_reconstruct_immutabele_updates) {
+  KeyValueBlockchain block_chain{db, true};
+
+  // Add block1
+  {
+    Updates updates;
+    ImmutableUpdates imm_updates;
+    imm_updates.addUpdate("imm_key1", ImmutableUpdates::ImmutableValue{"imm_value1", {"1", "2"}});
+    imm_updates.addUpdate("imm_key2", ImmutableUpdates::ImmutableValue{"imm_value2", {"12", "22"}});
+    updates.add("imm", std::move(imm_updates));
+
+    ASSERT_EQ(block_chain.addBlock(std::move(updates)), (BlockId)1);
+  }
+  const detail::Buffer& block_db_key = Block::generateKey(1);
+  auto block1_db_val = db->get(BLOCKS_CF, block_db_key);
+  ASSERT_TRUE(block1_db_val.has_value());
+
+  detail::Buffer in{block1_db_val.value().begin(), block1_db_val.value().end()};
+  auto block1_from_db = Block::deserialize(in);
+
+  detail::ImmutableKeyValueCategory cat{"imm", db};
+  auto wb = db->getBatch();
+  auto out = std::get<ImmutableOutput>(block1_from_db.data.categories_updates_info["imm"]);
+  cat.deleteBlock(out, wb);
+  db->write(std::move(wb));
+
+  ASSERT_THROW(categorization::RawBlock rw(block1_from_db, db), std::runtime_error);
```

Comment:
Can this ever happen?

Reply:
I hope not, but the test can be seen as to whether the raw block asserts when it fails to get any key that is present at the block.